src/condor_schedd.V6/qmgmt.cpp

```diff
@@ -1800,6 +1835,210 @@ void JobQueueJob::PopulateFromAd()
+		if (cad->set_id > 0) cad->Delete(ATTR_JOB_SET_ID);
+		cad->set_id = -1; // not in a jobset. remember that.
+	} else {
+		// verify the setid, and determine if we need to create a new jobset object
+		JobQueueJobSet * jobset = GetJobSetAd(cad->set_id);
+		if (jobset) {
+			// we might end up looking at the jobset ad before we iterate it, so make sure the ownerinfo is set
+			if ( ! jobset->ownerinfo) {
+				std::string owner2;
+				InitJobsetAd(jobset, owner2, is);
+				jobset_ids.push_back(jobset->Jobset());
+			}
+
+			// verify that the name of the jobset matches the name in the ad
+			jobset->LookupString(ATTR_JOB_SET_NAME, name2);
+			if (MATCH != strcasecmp(name1.c_str(), name2.c_str()) || jobset->ownerinfo != cad->ownerinfo) {
+				jobset = nullptr;
+			}
+		}
+
+		// we want a jobset, but don't know the id, so we add it to the map
+		// while also keeping track of the lowest cluster id that wants that setname/owner combo
+		if ( ! jobset) {
+			name2 = JobSets::makeAlias(name1, cad->ownerinfo->name);
+			unsigned int & setId = needed_sets[name2];
+			if (!setId || (setId > cad->jid.cluster)) { setId = cad->jid.cluster; }
+
+			// set set_id header field to 0 to mean 'figure out the id later'
+			// and if it was already set to non-zero,  that means the ad has a bogus attribute value
+			if (cad->set_id > 0) cad->Delete(ATTR_JOB_SET_ID);
+			cad->set_id = 0;
+		}
+	}
+	return true;
+}
+
+bool
+CreateNeededJobsets(
+	std::unordered_map<std::string, unsigned int> & needed_sets,  // map of jobset alias to lowest clusterid that needs that jobset
+	std::vector<unsigned int> & jobset_ids)
+{
+	int  fail_count = 0;
+	bool already_in_transaction = InTransaction();
+	if (!already_in_transaction) {
+		BeginTransaction();
+	}
+
+	// create the needed sets
+	for (auto it : needed_sets) {
+		const std::string & alias = it.first;
+		int cluster_id = it.second;
+		JobQueueCluster * cad = GetClusterAd(cluster_id);
+		if (! cad) {
+			// TODO: this is unexpected
+			dprintf(D_ERROR, "UNEXPECTED - no cluster ad for needed jobset id %d\n", cluster_id);
+			continue;
+		}
+		int jobset_id = cluster_id; // assume we will use the cluster id as the jobset id
+
+		// we don't expect there to be a jobset ad (yet), but if there is, we should check to see that the name is what we want.
+		JobQueueJobSet * sad = GetJobSetAd(jobset_id);
+		if (sad) {
+			std::string name2;
+			sad->LookupString(ATTR_JOB_SET_NAME, name2);
+			if (alias == JobSets::makeAlias(name2, sad->ownerinfo->name)) {
+				dprintf(D_FULLDEBUG, "need jobset id %d already has a jobset ad\n", jobset_id);
+			} else {
+				// allocate a new jobset id here
+				jobset_id = NewCluster();
+				dprintf(D_STATUS, "new jobset id %d allocated for %s\n", jobset_id, alias.c_str());
+				sad = nullptr;
+			}
+		}
+
+		std::string setname;
+		cad->LookupString(ATTR_JOB_SET_NAME, setname);
+		if ( ! JobSetCreate(jobset_id, setname.c_str(), cad->ownerinfo->Name())) {
+			++fail_count;
+			break;
```

Comment:
Should this be a continue?

Reply:
depends on the reason for the jobset create failure, I guess.  I'd be ok with continue on the theory that the failure might not be a general failure to create the new ad.

Reply:
Since it aborts on any failure and the count isn't returned, it probably doesn't matter. But it seems odd to have a count that doesn't go above 1.